src/hubbard/hubbard_utils.cpp

```diff
@@ -0,0 +1,215 @@
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <tuple>
+#include "context/simulation_context.hpp"
+
+namespace sirius {
+
+static std::vector<std::string>
+split(std::string str, std::string token)
+{
+    std::vector<std::string> result;
+    while (str.size()) {
+        int index = str.find(token);
+        if (index != std::string::npos) {
+            result.push_back(str.substr(0, index));
+            str = str.substr(index + token.size());
+            if (str.size() == 0)
+                result.push_back(str);
+        } else {
+            result.push_back(str);
+            str = "";
+        }
+    }
+    return result;
+}
+
+static auto
+parse_atom_string(std::string& at_str__) -> std::tuple<std::string, int, int>
+{
+    std::vector<std::string> split_str_ = split(at_str__, "_");
+    int n_                              = -1;
+    int l_                              = -1;
+
+    // a bit pedantic....
+    n_ = std::stoi(split_str_[1]);
+
+    switch (split_str_[1][1]) {
```

Comment:
there is a code repetition with https://github.com/electronic-structure/SIRIUS/blob/develop/src/unit_cell/atom_type.cpp#L412

better solution: introduce a pure function that returns l by charachter label. It can be defined for example in `atom_type.hpp`